packages/roosterjs-editor-api/lib/format/setIndentation.ts

```diff
@@ -18,32 +24,53 @@ import {
-    if (editor.isFeatureEnabled(ExperimentalFeatures.NewIndentation)) {
-        experimentSetIndentation(editor, indentation);
-    } else {
-        let command: DocumentCommand.Indent | DocumentCommand.Outdent =
-            indentation == Indentation.Increase ? DocumentCommand.Indent : DocumentCommand.Outdent;
-        editor.addUndoSnapshot(() => {
-            editor.focus();
-            let listNode = editor.getElementAtCursor('OL,UL');
-            let newNode: Node;
-
-            if (listNode) {
-                // There is already list node, setIndentation() will increase/decrease the list level,
-                // so we need to process the list when change indentation
-                newNode = processList(editor, command);
+    const handler = indentation == Indentation.Increase ? indent : outdent;
+
+    blockFormat(editor, (region, start, end) => {
+        const blocks = getSelectedBlockElementsInRegion(region, true /*createBlockIfEmpty*/);
+        const blockGroups: BlockElement[][] = [[]];
+
+        for (let i = 0; i < blocks.length; i++) {
+            const startNode = blocks[i].getStartNode();
+            const vList = createVListFromRegion(region, true /*includeSiblingLists*/, startNode);
+
+            if (vList) {
+                blockGroups.push([]);
+                while (blocks[i + 1] && vList.contains(blocks[i + 1].getStartNode())) {
+                    i++;
+                }
+                vList.setIndentation(start, end, indentation);
+                vList.writeBack();
             } else {
-                // No existing list node, browser will create <Blockquote> node for indentation.
-                // We need to set top and bottom margin to 0 to avoid unnecessary spaces
-                editor.getDocument().execCommand(command, false, null);
-                editor.queryElements('BLOCKQUOTE', QueryScope.OnSelection, node => {
-                    newNode = newNode || node;
-                    node.style.marginTop = '0px';
-                    node.style.marginBottom = '0px';
-                });
+                blockGroups[blockGroups.length - 1].push(blocks[i]);
             }
+        }
 
-            return newNode;
-        }, ChangeSource.Format);
-    }
+        blockGroups.forEach(group => handler(region, group));
+    });
+}
+
+function indent(region: RegionBase, blocks: BlockElement[]) {
+    const nodes = collapseNodesInRegion(region, blocks);
+    wrap(nodes, BlockWrapper);
+}
+
+function outdent(region: RegionBase, blocks: BlockElement[]) {
+    blocks.forEach(blockElement => {
+        let node = blockElement.collapseToSingleElement();
+        const quote = findClosestElementAncestor(node, region.rootNode, 'blockquote');
+        if (quote) {
+            if (node == quote) {
+                node = wrap(toArray(node.childNodes));
+            }
+
+            while (isNodeInRegion(region, node) && getTagOfNode(node) != 'BLOCKQUOTE') {

```

Comment:
Not sure if we can unify 'blockquote' and 'BLOCKQUOTE' to either upper case or lower case here?

Reply:
Good question. Ideally we should use the same case. However, for history reason, getTagOfNode() always returns upper case. And when search or create node, we prefer lower case so that typescript can infer the node type.

Reply:
okay, good to know that.